src/devices/src/virtio/block/device.rs

```diff
@@ -0,0 +1,948 @@
+
+        self.interrupt_evt.write(1).map_err(|e| {
+            error!("Failed to signal used queue: {:?}", e);
+            METRICS.block.event_fails.inc();
+            DeviceError::FailedSignalingUsedQueue(e)
+        })?;
+        Ok(())
+    }
+
+    /// Update the backing file for the Block device.
+    pub fn update_disk_image(&mut self, disk_image: File) -> result::Result<(), DeviceError> {
+        self.disk_image = disk_image;
+        self.disk_nsectors = self
+            .disk_image
+            .seek(SeekFrom::End(0))
+            .map_err(DeviceError::IoError)?
+            / SECTOR_SIZE;
+        self.disk_image_id = build_disk_image_id(&self.disk_image);
+        METRICS.block.update_count.inc();
+        Ok(())
+    }
+}
+
+impl VirtioDevice for Block {
+    fn device_type(&self) -> u32 {
+        TYPE_BLOCK
+    }
+
+    fn get_queues(&mut self) -> &mut Vec<Queue> {
+        &mut self.queues
+    }
+
+    fn get_queue_events(&self) -> Result<Vec<EventFd>, std::io::Error> {
+        Ok(vec![self.queue_evt.try_clone()?])
+    }
+
+    fn get_interrupt(&self) -> Result<EventFd, std::io::Error> {
+        Ok(self.interrupt_evt.try_clone()?)
+    }
+
+    /// Returns the current device interrupt status.
+    fn get_interrupt_status(&self) -> Arc<AtomicUsize> {
+        self.interrupt_status.clone()
+    }
+
+    fn avail_features(&self) -> u64 {
+        self.avail_features
+    }
+
+    fn acked_features(&self) -> u64 {
+        self.acked_features
+    }
+
+    fn set_acked_features(&mut self, acked_features: u64) {
+        self.acked_features = acked_features;
+    }
+
+    fn read_config(&self, offset: u64, mut data: &mut [u8]) {
+        let config_len = self.config_space.len() as u64;
+        if offset >= config_len {
+            error!("Failed to read config space");
+            METRICS.block.cfg_fails.inc();
+            return;
+        }
+        if let Some(end) = offset.checked_add(data.len() as u64) {
+            // This write can't fail, offset and end are checked against config_len.
+            data.write_all(&self.config_space[offset as usize..cmp::min(end, config_len) as usize])
+                .unwrap();
+        }
+    }
+
+    fn write_config(&mut self, offset: u64, data: &[u8]) {
+        let data_len = data.len() as u64;
+        let config_len = self.config_space.len() as u64;
+        if offset + data_len > config_len {
+            error!("Failed to write config space");
+            METRICS.block.cfg_fails.inc();
+            return;
+        }
+        let (_, right) = self.config_space.split_at_mut(offset as usize);
+        right.copy_from_slice(&data[..]);
+    }
+
+    fn is_activated(&self) -> bool {
+        self.device_activated
+    }
+
+    fn set_device_activated(&mut self, device_activated: bool) {
+        self.device_activated = device_activated;
+    }
+
+    fn activate(&mut self, _mem: GuestMemoryMmap) -> ActivateResult {
```

Comment:
Shouldn't the `fn activate`s have been removed from all devices?

Reply:
The `vsock` device still uses it. I've opened #1620 that tracks unifying the activation mechanism.